### PR TITLE
many: move resealing backend from boot to fdestate's backend

### PR DESCRIPTION
**Pull request by valentindavid**
We also make the FDE state manager install the the backend function to be associated with the state.

### CODE REVIEW

**Comment by valentindavid** [comment]
This depends on #14231 and #14234

**Review by pedronis — COMMENTED** [review]
some initial comments

File: overlord/fdestate/fdemgr.go [hunk 1]
```diff
@@ -22,6 +22,9 @@
 package fdestate
 
 import (
+	"github.com/snapcore/snapd/boot"
+	"github.com/snapcore/snapd/gadget/device"
+	"github.com/snapcore/snapd/overlord/fdestate/backend"
 	"github.com/snapcore/snapd/overlord/state"
 )
```
Comment (bboozzoo) [inline overlord/fdestate/fdemgr.go:30]: btw. just a line below which I can't comment on there's `managing managing`

File: overlord/fdestate/fdemgr.go [hunk 2]
```diff
@@ -38,6 +41,10 @@ func Manager(st *state.State, runner *state.TaskRunner) *FDEManager {
 		state: st,
 	}
 
+	boot.InstallResealKeyForBootChains(func(method device.SealingMethod, rootdir string, params *boot.ResealKeyForBootChainsParams, expectReseal bool) error {
+		return m.resealKeyForBootChains(method, rootdir, params, expectReseal)
+	})
```
Comment (bboozzoo) [inline overlord/fdestate/fdemgr.go:46]: this should work too
```suggestion
	boot.InstallResealKeyForBootChains(m.resealKeyForBootChains)
```

File: boot/seal.go [hunk 3]
```diff
@@ -442,11 +429,11 @@ func resealKeyToModeenvImpl(rootdir string, modeenv *Modeenv, expectReseal bool,
 	return resealKeyToModeenvForMethod(method, rootdir, modeenv, expectReseal)
 }
 
-type resealKeyForBootChainsParams struct {
-	runModeBootChains           []bootChain
-	recoveryBootChainsForRunKey []bootChain
-	recoveryBootChains          []bootChain
-	roleToBlName                map[bootloader.Role]string
+type ResealKeyForBootChainsParams struct {
+	RunModeBootChains           []bootChain
```
Comment (bboozzoo) [inline boot/seal.go:433]: do you need s/bootChain/BootChain/ too?

File: boot/seal.go [hunk 4]
```diff
@@ -531,74 +518,17 @@ func resealKeyToModeenvForMethod(method device.SealingMethod, rootdir string, mo
...
-
-	return nil
+func resealKeyForBootChainsImpl(method device.SealingMethod, rootdir string, params *ResealKeyForBootChainsParams, expectReseal bool) error {
+	return fmt.Errorf("FDE manager was not started")
 }
 
 var resealKeyForBootChains = resealKeyForBootChainsImpl
 
-func MockResealKeyForBootChains(f func(method device.SealingMethod, rootdir string, params *resealKeyForBootChainsParams, expectReseal bool) error) (restore func()) {
+func InstallResealKeyForBootChains(f func(method device.SealingMethod, rootdir string, params *ResealKeyForBootChainsParams, expectReseal bool) error) {
```
Comment (pedronis) [inline boot/seal.go:527]: we usually just have a top level variable with a func value without an install function fwiw

File: boot/seal.go (hunk 3 above)
Comment (valentindavid) [reply boot/seal.go:433]: The capitalization of some of those names are only temporary until everything is moved to `overlord/fdestate/backend`. So I will see later what I need to make public. But maybe `bootChain` will move. Anyway, I do the minimum here.

File: overlord/fdestate/fdemgr.go (hunk 1 above)
Comment (valentindavid) [reply overlord/fdestate/fdemgr.go:30]: I will open a mini PR for that.

Comment (valentindavid) [reply overlord/fdestate/fdemgr.go:30]: #14246

File: boot/seal.go (hunk 3 above)
Comment (valentindavid) [reply boot/seal.go:433]: Done along with @pedronis related comment.

File: overlord/fdestate/fdemgr.go (hunk 2 above)
Comment (valentindavid) [reply overlord/fdestate/fdemgr.go]: Done along with related comment from @pedronis

File: overlord/fdestate/backend/reseal.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+func MockSecbootResealKeys(f func(params *secboot.ResealKeysParams) error) (restore func()) {
+	osutil.MustBeTestBinary("secbootResealKeys only can be mocked in tests")
+	old := secbootResealKeys
+	secbootResealKeys = f
+	return func() {
+		secbootResealKeys = old
+	}
+}
+
+// ResealKeyForBootChains reseals disk encryption keys with the given bootchains
```
Comment (pedronis) [inline overlord/fdestate/backend/reseal.go:51]: missing . at the end

**Review by pedronis — APPROVED** [review]
this looks reasonable as an intermediate step, as discussed more things, especially some that now have been made public will move to the fde mgr backend once we move seal code as well

**Review by bboozzoo — APPROVED** [review]
LGTM

**Pull request merged**